lib/mps_simulator.h

```diff
@@ -0,0 +1,170 @@
+// Copyright 2019 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MPS_SIMULATOR_H_
+#define MPS_SIMULATOR_H_
+
+// For templates will take care of parallelization.
+#define EIGEN_DONT_PARALLELIZE 1
+
+#include <cstdint>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <vector>
+
+#include "../eigen/Eigen/Dense"
+#include "mps_statespace.h"
+
+namespace qsim {
+
+namespace mps {
+
+/**
+ * Quantum circuit simulator without vectorization.
+ */
+template <typename For, typename fp_type = float>
+class MPSSimulator final {
+ public:
+  using MPSStateSpace_ = MPSStateSpace<For, fp_type>;
+  using MPS = typename MPSStateSpace_::MPS;
+
+  using Matrix = Eigen::Matrix<std::complex<fp_type>, Eigen::Dynamic,
+                               Eigen::Dynamic, Eigen::RowMajor>;
+  using ConstMatrixMap = Eigen::Map<const Matrix>;
+  using MatrixMap = Eigen::Map<Matrix>;
+
+  using OneQBMatrix =
+      Eigen::Matrix<std::complex<fp_type>, 2, 2, Eigen::RowMajor>;
+  using ConstOneQBMap = Eigen::Map<const OneQBMatrix>;
+
+  // Note: ForArgs are currently unused.
+  template <typename... ForArgs>
+  explicit MPSSimulator(ForArgs&&... args) : for_(args...) {}
+
+  /**
+   * Applies a gate using non-vectorized instructions.
+   * @param qs Indices of the qubits affected by this gate.
+   * @param matrix Matrix representation of the gate to be applied.
+   * @param state The state of the system, to be updated by this method.
+   */
+  void ApplyGate(const std::vector<unsigned>& qs, const fp_type* matrix,
+                 MPS& state) const {
+    // Assume qs[0] < qs[1] < qs[2] < ... .
+
+    switch (qs.size()) {
+      case 1:
+        ApplyGate1(qs, matrix, state);
+        break;
+      // case 2:
+      //   ApplyGate2(qs, matrix, state);
+      //   break;
+      // case 3:
+      //   ApplyGate3(qs, matrix, state);
+      //   break;
+      // case 4:
+      //   ApplyGate4(qs, matrix, state);
+      //   break;
+      // case 5:
+      //   ApplyGate5(qs, matrix, state);
+      //   break;
+      // case 6:
+      //   ApplyGate6(qs, matrix, state);
+      //   break;
+      default:
+        // Not implemented.
+        break;
+    }
+  }
+
```

Comment:
Make the following three functions private?

Reply:
Oops. Done.